diff --git a/python/mozbuild/mozbuild/frontend/emitter.py b/python/mozbuild/mozbuild/frontend/emitter.py
index 5d2857089d..e18e930a92 100644
--- a/python/mozbuild/mozbuild/frontend/emitter.py
+++ b/python/mozbuild/mozbuild/frontend/emitter.py
@@ -1345,7 +1345,6 @@ class TreeMetadataEmitter(LoggingMixin):
                 Defines,
                 (
                     computed_flags,
-                    computed_as_flags,
                 ),
             ),
             ("HOST_DEFINES", HostDefines, (computed_host_flags,)),
