commit 500754d844368b01d51e26e05691370bc5aaa632
Author: Josh Matthews <josh@joshmatthews.net>
Date:   Wed Feb 3 23:42:28 2021 -0500

    remove unused windows code

diff --git a/mozglue/misc/moz.build b/mozglue/misc/moz.build
index 96e468b2d..2be17cb0a 100644
--- a/mozglue/misc/moz.build
+++ b/mozglue/misc/moz.build
@@ -30,7 +30,6 @@ EXPORTS.mozilla.glue += [
 if CONFIG["OS_ARCH"] == "WINNT":
     EXPORTS.mozilla += [
         "GetKnownFolderPath.h",
-        "PreXULSkeletonUI.h",
         "StackWalk_windows.h",
         "StackWalkThread.h",
         "TimeStamp_windows.h",
@@ -82,7 +81,6 @@ if CONFIG["OS_ARCH"] == "WINNT":
         SOURCES += [
             "/ipc/mscom/COMWrappers.cpp",
             "/ipc/mscom/ProcessRuntime.cpp",
-            "PreXULSkeletonUI.cpp",
         ]
 
     if CONFIG["ACCESSIBILITY"]:
