From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Anthony Ramine <nox@nox.paris>
Date: Tue, 7 Jan 2020 12:16:05 +0100
Subject: [PATCH] Do not rely on internal .cargo/config


diff --git a/config/recurse.mk b/config/recurse.mk
index 7956578ac..7ff2ec209 100644
--- a/config/recurse.mk
+++ b/config/recurse.mk
@@ -201,24 +201,6 @@ endif
 # Those need to depend on config/export for system wrappers.
 $(addprefix build/unix/stdc++compat/,target host) build/clang-plugin/host: config/export

-# Rust targets, and export targets that run cbindgen need
-# $topobjdir/.cargo/config.toml to be preprocessed first. Ideally, we'd only set it
-# as a dependency of the rust targets, but unfortunately, that pushes Make to
-# execute them much later than we'd like them to be when the file doesn't exist
-# prior to Make running. So we also set it as a dependency of pre-export, which
-# ensures it exists before recursing the rust targets and the export targets
-# that run cbindgen, tricking Make into keeping them early.
-# When $topobjdir/.cargo/config exists from an old build, we also remove it because
-# cargo will prefer to use it rather than config.toml.
-CARGO_CONFIG_DEPS = $(DEPTH)/.cargo/config.toml
-ifneq (,$(wildcard $(DEPTH)/.cargo/config))
-CARGO_CONFIG_DEPS += $(MDDEPDIR)/cargo-config-cleanup.stub
-endif
-$(rust_targets): $(CARGO_CONFIG_DEPS)
-ifndef TEST_MOZBUILD
-recurse_pre-export: $(CARGO_CONFIG_DEPS)
-endif
-
 $(MDDEPDIR)/cargo-config-cleanup.stub:
 	rm $(DEPTH)/.cargo/config
 	touch $@
diff --git a/moz.build b/moz.build
index 9db27fd05..a330fd1aa 100644
--- a/moz.build
+++ b/moz.build
@@ -183,10 +183,6 @@ if CONFIG["MOZ_BUILD_APP"]:
 else:
     include("/toolkit/toolkit.mozbuild")
 
-OBJDIR_PP_FILES[".cargo"] += [
-    CONFIG["MOZ_OVERRIDE_CARGO_CONFIG"] or ".cargo/config.toml.in"
-]
-
 DEFINES["top_srcdir"] = TOPSRCDIR
 
 SPHINX_TREES["contributing"] = "docs/contributing"
