diff --git a/build/moz.configure/toolchain.configure b/build/moz.configure/toolchain.configure
index 8a97d952d..5b538f0a1 100755
--- a/build/moz.configure/toolchain.configure
+++ b/build/moz.configure/toolchain.configure
@@ -100,7 +100,7 @@ with only_when(host_is_osx | target_is_osx):
     )
 
     def mac_sdk_min_version():
-        return "14.4"
+        return "14.2"
 
     @depends(
         "--with-macos-sdk",
