diff --git a/config/external/zlib/moz.build b/config/external/zlib/moz.build
index 755400fa7..d1311b9a7 100644
--- a/config/external/zlib/moz.build
+++ b/config/external/zlib/moz.build
@@ -6,16 +6,4 @@
 
 Library("zlib")
 
-if CONFIG["MOZ_SYSTEM_ZLIB"]:
-    OS_LIBS += CONFIG["MOZ_ZLIB_LIBS"]
-else:
-    if CONFIG["ZLIB_IN_MOZGLUE"]:
-        # Can't do this until mozglue is handled by moz.build instead of
-        # config/rules.mk.
-        # USE_LIBS += [
-        #     'mozglue'
-        # ]
-        pass
-    DIRS += [
-        "../../../modules/zlib",
-    ]
+OS_LIBS += CONFIG["MOZ_ZLIB_LIBS"]
diff --git a/moz.configure b/moz.configure
index 804b9a375..87ccc508d 100755
--- a/moz.configure
+++ b/moz.configure
@@ -905,52 +905,11 @@ check_prog(
     validate=validate_strip,
 )
 
+pkg_check_modules("MOZ_ZLIB", "zlib >= 1.2.3", when=use_pkg_config)
 
-@depends(js_standalone, target)
-def system_zlib_default(js_standalone, target):
-    return (
-        js_standalone
-        and target.kernel not in ("WINNT", "Darwin")
-        and target.os != "Android"
-    )
-
-
-option(
-    "--with-system-zlib",
-    nargs="?",
-    default=system_zlib_default,
-    help="{Use|Do not use} system libz",
-    when=use_pkg_config,
-)
-
-
-@depends("--with-system-zlib", when=use_pkg_config)
-def with_system_zlib_option(with_system_zlib):
-    return with_system_zlib
-
-
-@depends(with_system_zlib_option)
-def deprecated_system_zlib_path(value):
-    if value and len(value) == 1:
-        die(
-            "--with-system-zlib=PATH is not supported anymore. Please use "
-            "--with-system-zlib and set any necessary pkg-config environment variable."
-        )
-
-
-pkg_check_modules("MOZ_ZLIB", "zlib >= 1.2.3", when="--with-system-zlib")
-
-set_config("MOZ_SYSTEM_ZLIB", True, when="--with-system-zlib")
-
-
-@depends(with_system_zlib_option, js_shared, moz_linker, target.os)
-def zlib_in_mozglue(system_zlib, js_shared, linker, os):
-    if not system_zlib and (js_shared or linker or os == "Android"):
-        return True
-
-
-set_config("ZLIB_IN_MOZGLUE", zlib_in_mozglue)
-set_define("ZLIB_IN_MOZGLUE", zlib_in_mozglue)
+set_config("MOZ_SYSTEM_ZLIB", True)
+# This is not in upstream anymore
+set_config("ZLIB_IN_MOZGLUE", False)
 
 
 # Please do not add configure checks from here on.
