diff --git a/config/external/zlib/moz.build b/config/external/zlib/moz.build
index 755400fa7..d1311b9a7 100644
--- a/config/external/zlib/moz.build
+++ b/config/external/zlib/moz.build
@@ -6,9 +6,4 @@
 
 Library("zlib")
 
-if CONFIG["MOZ_SYSTEM_ZLIB"]:
-    OS_LIBS += CONFIG["MOZ_ZLIB_LIBS"]
-else:
-    DIRS += [
-        "../../../modules/zlib",
-    ]
+OS_LIBS += CONFIG["MOZ_ZLIB_LIBS"]
diff --git a/moz.configure b/moz.configure
index 804b9a375..87ccc508d 100755
--- a/moz.configure
+++ b/moz.configure
@@ -891,50 +891,8 @@ check_prog(
     validate=validate_strip,
 )
 
+pkg_check_modules("MOZ_ZLIB", "zlib >= 1.2.3", when=use_pkg_config)
 
-@depends(js_standalone, target)
-def system_zlib_default(js_standalone, target):
-    return (
-        js_standalone
-        and target.kernel not in ("WINNT", "Darwin")
-        and target.os != "Android"
-    )
-
-
-option(
-    "--with-system-zlib",
-    nargs="?",
-    default=system_zlib_default,
-    help="{Use|Do not use} system libz",
-    when=use_pkg_config,
-)
-
-
-@depends("--with-system-zlib", when=use_pkg_config)
-def with_system_zlib_option(with_system_zlib):
-    return with_system_zlib
-
-
-@depends(with_system_zlib_option)
-def deprecated_system_zlib_path(value):
-    if value and len(value) == 1:
-        die(
-            "--with-system-zlib=PATH is not supported anymore. Please use "
-            "--with-system-zlib and set any necessary pkg-config environment variable."
-        )
-
-
-pkg_check_modules("MOZ_ZLIB", "zlib >= 1.2.3", when="--with-system-zlib")
-
-set_config("MOZ_SYSTEM_ZLIB", True, when="--with-system-zlib")
-
-option(
-    env="USE_LIBZ_RS",
-    default=milestone.is_early_beta_or_earlier,
-    help="Use libz-rs-sys instead of zlib",
-    when=toolkit & ~with_system_zlib_option,
-)
-
-set_config("USE_LIBZ_RS", True, when="USE_LIBZ_RS")
+set_config("MOZ_SYSTEM_ZLIB", True)
 
 with only_when(cross_compiling):
     option(