diff --git a/js/src/jit/ProcessExecutableMemory.cpp b/js/src/jit/ProcessExecutableMemory.cpp
--- a/js/src/jit/ProcessExecutableMemory.cpp	(revision 5ef22dcf66d5aa396ad6fe8dd53dac88a29f8774)
+++ b/js/src/jit/ProcessExecutableMemory.cpp	(revision c9a9b9b6b8f6aba9858ca14171f8c0051baa272a)
@@ -584,9 +584,15 @@
 #    endif
   return true;
 #  else
-  unsigned flags = ProtectionSettingToFlags(protection);
-  void* p = MozTaggedAnonymousMmap(addr, bytes, flags,
-                                   MAP_FIXED | MAP_PRIVATE | MAP_ANON, -1, 0,
+  unsigned prot_flags = ProtectionSettingToFlags(protection);
+  int flags = MAP_FIXED | MAP_PRIVATE | MAP_ANON;
+#    ifdef XP_OHOS
+  // Required for JIT code on HarmonyOS.
+  // Since MAP_EXECUTABLE is documented to be ignored on Linux, we
+  // unconditionally enable it for all OpenHarmony distributions.
+  flags |= MAP_EXECUTABLE;
+#    endif
+  void* p = MozTaggedAnonymousMmap(addr, bytes, prot_flags, flags, -1, 0,
                                    "js-executable-memory");
   if (p == MAP_FAILED) {
     return false;
