diff --git a/intl/components/src/Collator.h b/intl/components/src/Collator.h
index 27a59188f..1396c9086 100644
--- a/intl/components/src/Collator.h
+++ b/intl/components/src/Collator.h
@@ -10,7 +10,7 @@
 
 #include <type_traits>
 
-#include "mozilla/intl/collator_glue.h"
+#include "collator_glue.h"
 
 #include "mozilla/intl/ICU4CGlue.h"
 #include "mozilla/intl/ICUError.h"
diff --git a/intl/components/src/Bidi.h b/intl/components/src/Bidi.h
index 8db74173e..2e8ffa317 100644
--- a/intl/components/src/Bidi.h
+++ b/intl/components/src/Bidi.h
@@ -13,7 +13,7 @@
 #define USE_RUST_UNICODE_BIDI 1
 
 #if USE_RUST_UNICODE_BIDI
-#  include "mozilla/intl/unicode_bidi_ffi_generated.h"
+#  include "unicode_bidi_ffi_generated.h"
 #else
 struct UBiDi;
 #endif
diff --git a/js/src/builtin/String.cpp b/js/src/builtin/String.cpp
index 150c0f65f..77d57a3e5 100644
--- a/js/src/builtin/String.cpp
+++ b/js/src/builtin/String.cpp
@@ -42,7 +42,7 @@
 #if !JS_HAS_INTL_API
 #  include "js/LocaleSensitive.h"
 #endif
-#include "js/normalizer_glue.h"
+#include "normalizer_glue.h"
 #include "js/Prefs.h"
 #include "js/Printer.h"
 #include "js/PropertyAndElement.h"  // JS_DefineFunctions
diff --git a/js/src/builtin/intl/Locale.cpp b/js/src/builtin/intl/Locale.cpp
index bd1a1e37b..453921951 100644
--- a/js/src/builtin/intl/Locale.cpp
+++ b/js/src/builtin/intl/Locale.cpp
@@ -28,7 +28,7 @@
 #include "builtin/Boolean.h"
 #include "builtin/intl/CommonFunctions.h"
 #include "builtin/intl/FormatBuffer.h"
-#include "builtin/intl/glue/Locale.h"
+#include "locale_glue.h"
 #include "builtin/intl/LanguageTag.h"
 #include "builtin/intl/LocaleNegotiation.h"
 #include "builtin/intl/ParameterNegotiation.h"
diff --git a/js/src/irregexp/imported/regexp-parser.cc b/js/src/irregexp/imported/regexp-parser.cc
index 3926339e9..6df564ece 100644
--- a/js/src/irregexp/imported/regexp-parser.cc
+++ b/js/src/irregexp/imported/regexp-parser.cc
@@ -10,7 +10,7 @@
 #include "irregexp/imported/regexp.h"
 
 #ifdef V8_INTL_SUPPORT
-#include "js/properties_glue.h"
+#include "properties_glue.h"
 #include "unicode/uniset.h"
 #include "unicode/unistr.h"
 #include "unicode/usetiter.h"
diff --git a/js/src/moz.build b/js/src/moz.build
index 0a62964f3..2e0ab1d02 100755
--- a/js/src/moz.build
+++ b/js/src/moz.build
@@ -309,7 +309,6 @@ DIRS += [
     "vm",
     "wasm",
     "xsum",
-    "builtin/normalizer_glue",
 ]
 
 # Glean integration
@@ -339,9 +338,6 @@ if CONFIG["JS_HAS_INTL_API"]:
     DIRS += [
         "builtin/intl",
         "builtin/temporal",
-        "builtin/intl/collator_glue",
-        "builtin/intl/locale_glue",
-        "irregexp/properties_glue",
     ]
 
 CXXFLAGS += CONFIG["MOZ_ZLIB_CFLAGS"]
@@ -367,7 +363,6 @@ GeneratedFile(
 if CONFIG["JS_HAS_INTL_API"]:
     if CONFIG["JS_STANDALONE"]:
         DIRS += [
-            "../../intl/bidi",
             "../../intl/components",
         ]
 
