diff --git a/js/public/MemoryMetrics.h b/js/public/MemoryMetrics.h
index 69ecf27d4..c38eb9a09 100644
--- a/js/public/MemoryMetrics.h
+++ b/js/public/MemoryMetrics.h
@@ -941,6 +941,11 @@ extern JS_PUBLIC_API bool AddSizeOfTab(JSContext* cx, JS::Zone* zone,
                                        TabSizes* sizes,
                                        const JS::AutoRequireNoGC& nogc);
 
+extern JS_PUBLIC_API bool AddServoSizeOf(JSContext* cx,
+                                         mozilla::MallocSizeOf mallocSizeOf,
+                                         ObjectPrivateVisitor* opv,
+                                         ServoSizes* sizes);
+
 }  // namespace JS
 
 #undef DECL_SIZE_ZERO
diff --git a/js/src/vm/MemoryMetrics.cpp b/js/src/vm/MemoryMetrics.cpp
index 5c3e7b6e5..f79dc06e8 100644
--- a/js/src/vm/MemoryMetrics.cpp
+++ b/js/src/vm/MemoryMetrics.cpp
@@ -895,4 +895,32 @@ JS_PUBLIC_API bool AddSizeOfTab(JSContext* cx, JS::Zone* zone,
   return true;
 }
 
+JS_PUBLIC_API bool AddServoSizeOf(JSContext* cx, MallocSizeOf mallocSizeOf,
+                                  ObjectPrivateVisitor* opv,
+                                  ServoSizes* sizes) {
+  SimpleJSRuntimeStats rtStats(mallocSizeOf);
+
+  // No need to anonymize because the results will be aggregated.
+  if (!CollectRuntimeStatsHelper(cx, &rtStats, opv, /* anonymize = */ false,
+                                 StatsCellCallback<CoarseGrained>))
+    return false;
+
+#ifdef DEBUG
+  size_t gcHeapTotalOriginal = sizes->gcHeapUsed + sizes->gcHeapUnused +
+                               sizes->gcHeapAdmin + sizes->gcHeapDecommitted;
+#endif
+
+  rtStats.addToServoSizes(sizes);
+  rtStats.zTotals.addToServoSizes(sizes);
+  rtStats.realmTotals.addToServoSizes(sizes);
+
+#ifdef DEBUG
+  size_t gcHeapTotal = sizes->gcHeapUsed + sizes->gcHeapUnused +
+                       sizes->gcHeapAdmin + sizes->gcHeapDecommitted;
+  MOZ_ASSERT(rtStats.gcHeapChunkTotal == gcHeapTotal - gcHeapTotalOriginal);
+#endif
+
+  return true;
+}
+
 }  // namespace JS
