diff --git a/configure.py b/configure.py
index b8b217f48..62495a069 100644
--- a/configure.py
+++ b/configure.py
@@ -107,7 +107,7 @@ def main(argv):
             # 2. Long source paths repeated across many -I flags can exceed
             #    the CreateProcessW command line limit of 32,767 characters.
             WIN32_MAX_PATH = 260
-            LONGEST_KNOWN_OBJDIR_RELATIVE_PATH = 170
+            LONGEST_KNOWN_OBJDIR_RELATIVE_PATH = 100
             DEFAULT_OBJDIR_NAME_LEN = 28  # /obj-x86_64-pc-windows-msvc/
             max_objdir_len = WIN32_MAX_PATH - LONGEST_KNOWN_OBJDIR_RELATIVE_PATH
             # Account for the default objdir name in the srcdir limit so that
diff --git a/python/mozboot/bin/bootstrap.py b/python/mozboot/bin/bootstrap.py
index 12ceabc38..877212aa6 100755
--- a/python/mozboot/bin/bootstrap.py
+++ b/python/mozboot/bin/bootstrap.py
@@ -74,7 +74,7 @@ def validate_clone_dest(dest: Path):
     if WINDOWS:
         # Keep in sync with the path length checks in configure.py.
         WIN32_MAX_PATH = 260
-        LONGEST_KNOWN_OBJDIR_RELATIVE_PATH = 170
+        LONGEST_KNOWN_OBJDIR_RELATIVE_PATH = 100
         DEFAULT_OBJDIR_NAME_LEN = 28  # /obj-x86_64-pc-windows-msvc/
         max_srcdir_len = (
             WIN32_MAX_PATH
