diff --git a/js/public/Promise.h b/js/public/Promise.h
index ffef2095ad..1bc86acc70 100644
--- a/js/public/Promise.h
+++ b/js/public/Promise.h
@@ -155,6 +155,8 @@ class JS_PUBLIC_API JobQueue {
  */
 extern JS_PUBLIC_API void SetJobQueue(JSContext* cx, JobQueue* queue);
 
+extern JS_PUBLIC_API JobQueue* GetJobQueue(JSContext* cx);
+
 /**
  * [SMDOC] Protecting the debuggee's job/microtask queue from debugger activity.
  *
diff --git a/js/src/jsapi.cpp b/js/src/jsapi.cpp
index 5fafa677de..805809ada2 100644
--- a/js/src/jsapi.cpp
+++ b/js/src/jsapi.cpp
@@ -2862,6 +2862,11 @@ JS_PUBLIC_API void JS::SetJobQueue(JSContext* cx, JobQueue* queue) {
   cx->jobQueue = queue;
 }
 
+JS_PUBLIC_API JS::JobQueue* JS::GetJobQueue(JSContext* cx) {
+  MOZ_ASSERT(cx->jobQueue.ref());
+  return cx->jobQueue;
+}
+
 extern JS_PUBLIC_API void JS::SetPromiseRejectionTrackerCallback(
     JSContext* cx, PromiseRejectionTrackerCallback callback,
     void* data /* = nullptr */) {
