diff --git a/moz.configure b/moz.configure
index 3a9461200..9cf45c632 100755
--- a/moz.configure
+++ b/moz.configure
@@ -697,12 +697,17 @@ check_prog(

 
 @template
-def plain_llvm_or_prefixed(name):
-    @depends(llvm_tool(f"llvm-{name}"), toolchain_prefix)
+def plain_llvm_or_prefixed(name, llvm_name=None):
+    # look for a tool, using the following alternatives, in that order:
+    # 1. llvm-${llvm_name}, or llvm-${name} if ${llvm_name} is not provided
+    # 2. ${toolchain_prefix}${name}
+    # 3. ${name}
+
+    @depends(llvm_tool("llvm-{}".format(llvm_name or name)), toolchain_prefix)
     def plain_llvm_or_prefixed(llvm_tool, toolchain_prefix):
         commands = [llvm_tool[0], name]
         for prefix in toolchain_prefix or ():
-            commands.insert(1, f"{prefix}{name}")
+            commands.insert(1, "{}{}".format(prefix, name))
         return tuple(commands)
 
     return plain_llvm_or_prefixed
@@ -717,7 +722,7 @@ check_prog(
 
 check_prog(
     "INSTALL_NAME_TOOL",
-    plain_llvm_or_prefixed("install-name-tool"),
+    plain_llvm_or_prefixed("install_name_tool", llvm_name="install-name-tool"),
     when=compile_environment & target_is_darwin & js_standalone,
     paths=clang_search_path,
 )
