diff --git a/js/src/irregexp/imported/regexp-graph-printer.cc b/js/src/irregexp/imported/regexp-graph-printer.cc
--- a/imported/regexp-graph-printer.cc
+++ b/imported/regexp-graph-printer.cc
@@ -7,6 +7,8 @@
 #include "irregexp/imported/regexp-graph-printer.h"

 #include <iomanip>
+#include <queue>
+#include <unordered_set>

 #include "irregexp/imported/regexp-ast-printer.h"
 #include "irregexp/imported/regexp-compiler.h"
